src/consensus/consensus.md

```diff
@@ -0,0 +1,293 @@
+validity of blocks, the Consensus layer can remain oblivious to the details on validation
+and rely on the Ledger layer to make such judgment, based on the particular era in effect.
+
+## Selecting chains
+
+_Chain selection_ is the mechanism, specified in the Protocol associated to the current
+era in effect in the network, through which nodes acquire and diffuse the best chain in
+the network. It is important to note that **all nodes** participating in the network are in
+charge of this responsibility, as opposed to producing blocks, which only some nodes will
+be involved in doing (namely those with forging credentials, the cryptographic keys for
+signing new blocks on behalf of a stake pool).
+
+### Header|body split
+
+An essential and uncontroversial design refinement in any blockchain implementation is to
+separate block headers and block bodies:
+
+- If blocks can be almost fully validated in constant time based on looking at only a small fixed
+size block header, then honest nodes can validate candidate chains with a small bounded
+amount of work.
+
+- It also enables a design where a node can see blocks available from many immediate
+peers but can choose to download each block body of interest just once (from a peer of its
+choosing from which it is available). This saves network bandwidth.
+
+In the case of Ouroboros, all the cryptographic consensus evidence is packed into the
+block header, leaving the block body containing only the ledger data, and check that the
+block has been signed by a node that is the slot leader. If we validate this in the
+context of a chain of headers, then we can establish this is a plausible candidate
+chain, thus we eliminate several potential resource draining attacks.
+
+So the design at this stage involves transmitting chains of headers rather than whole
+blocks, and using a secondary mechanism to download block bodies of interest.
+
+The Consensus chain selection is involved with chains of block headers, whereas the
+validity check is performed by the Ledger rules, effectively separating concerns.
+
+
+> [!NOTE]
+> Chain selection is performed based on data from the previous headers, so there is some state
+> (specific to each protocol) that influences the selection, which we will call _chain
+> state_, as opposed to the state of the ledger, which we will call _ledger state_.
+
+### The `k` security parameter
+
+Both Ouroboros Classic and Ouroboros Praos are based on a chain
+selection rule that imposes a maximum rollback condition: alternative
+chains to a node's current chain that fork off more than a certain
+number of blocks ago are never considered for adoption. This limit is
+known as the security parameter, and is usually denoted by `k`, which
+on mainnet is set to 2160 blocks. The analysis shows that nodes will
+not diverge more than `k` blocks under honest circumstances.
+
+Ouroboros BFT does not impose a maximum rollback, but adding such a
```

Comment:
Do we care about BFT at this stage?

Reply:
I think we care about it at least for implementation reasons, like replaying the chain needs to know how to process blocks from BFT.